core/src/main/java/org/apache/calcite/rel/rules/AggregateReduceFunctionsRule.java

```diff
@@ -97,12 +98,96 @@
       new AggregateReduceFunctionsRule(operand(LogicalAggregate.class, any()),
           RelFactories.LOGICAL_BUILDER);
 
+  private final AggregateFunctionsToReduce functionsToReduce;
+
+  /**
+   * Used to create a non-default version of this rule
+   * where we can be specific about which functions
+   * should be reduced by the rule
+   */
+  public static class AggregateFunctionsToReduce {
+    private final boolean reduceAvg;
+    private final boolean reduceSum;
+    private final boolean reduceStddevPop;
+    private final boolean reduceStddevSamp;
+    private final boolean reduceVarPop;
+    private final boolean reduceVarSamp;
+    private final boolean reduceCovarPop;
+    private final boolean reduceCovarSamp;
+    private final boolean reduceRegrSxx;
+    private final boolean reduceRegrSyy;
+
+    private AggregateFunctionsToReduce() {
+      this.reduceAvg = true;
+      this.reduceSum = true;
+      this.reduceStddevPop = true;
+      this.reduceStddevSamp = true;
+      this.reduceVarPop = true;
+      this.reduceVarSamp = true;
+      this.reduceCovarPop = true;
+      this.reduceCovarSamp = true;
+      this.reduceRegrSxx = true;
+      this.reduceRegrSyy = true;
+    }
+
+    public AggregateFunctionsToReduce(final boolean reduceAvg, final boolean reduceSum,
+        final boolean reduceStddevPop, final boolean reduceStddevSamp,
+        final boolean reduceVarPop, final boolean reduceVarSamp,
+        final boolean reduceCovarPop, final boolean reduceCovarSamp,
+        final boolean reduceRegrSxx, final boolean reduceRegrSyy) {
+      this.reduceAvg = reduceAvg;
+      this.reduceSum = reduceSum;
+      this.reduceStddevPop = reduceStddevPop;
+      this.reduceStddevSamp = reduceStddevSamp;
+      this.reduceVarPop = reduceVarPop;
+      this.reduceVarSamp = reduceVarSamp;
+      this.reduceCovarPop = reduceCovarPop;
+      this.reduceCovarSamp = reduceCovarSamp;
+      this.reduceRegrSxx = reduceRegrSxx;
+      this.reduceRegrSyy = reduceRegrSyy;
+    }
+  }
+
+  /**
+   * Gets an instance of AggregateReduceFunctionsRule
+   * with client provided specific functions to reduce
+   * @param functionsToReduce client provided information
+   *                          on which specific functions will be
+   *                          reduced by this rule
+   * @return an instance of AggregateReduceFunctionsRule
+   */
+  public static AggregateReduceFunctionsRule
+      getInstanceWithSpecificFunctionsToReduce(final AggregateFunctionsToReduce functionsToReduce) {
```

Comment:
Having both a public constructor and a public factory method seems a bit redundant. In general, I like static factories but since the existing rule base does not have many such factories, I would prefer to keep only the constructor mostly for homogeneity. What do you think?

Reply:
The reason why I used a static factory method is because I wasn't sure if the callers will always be able to create a constructor themselves if they don't have knowledge of the arguments RelOptRuleOperand and RelBuilderFactory. 

Until now the callers always grab instance as AggregateReduceFunctionsRule.INSTANCE and they will continue to do so to get the default instance of this rule that reduces all functions.

For callers interested in reducing specific functions, they can call the factory method with Set<SqlKind> functionsToReduce. 

I am just not sure if the knowledge of the constructor's arguments is there in the client. Like do they always know what to pass in for RelOptRuleOperand and RelBuilderFactory. So I thought public factory method is the way to go. Is that fine?

Reply:
You raised a valid point which remind me a few more stuff to take into account in the constructor. I will add more comments.  

However, I'm still not very much in favor of the static factory. It creates a very specialized rule that applies only for LogicalAggregates and uses a factory of Logical expressions which I am not sure if it corresponds to the most common use case. Moreover, helping the client use the constructor could also be part of the Javadoc.

Reply:
Done.